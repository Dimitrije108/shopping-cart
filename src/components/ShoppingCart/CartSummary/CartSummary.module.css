.cartSummaryCont {
	padding-left: 2rem;
	border-left: 1px solid rgb(104, 104, 104);

	display: flex;
	flex-direction: column;
	justify-content: flex-start;

	align-self: start;

	position: sticky;
	top: 4.5rem;
}

.summaryHeader {
	margin: 0;
	padding: 0;
	margin-bottom: 0.5rem;
	padding-bottom: 0.5rem;

	font-family: 'DINNext bold';
	border-bottom: 2px solid rgb(104, 104, 104);
}

.wrapper,
.discountLayout,
.totalLayout {
	display: flex;
	justify-content: space-between;
	align-items: center;
}

.discountWrapper {
	padding: 1rem 0;
	border-bottom: 1px solid rgb(104, 104, 104);
}

.discount {
	margin-right: 0.5rem;
	flex: 1;
}

.codeInput {
	padding: 2px 4px;

	background-color:rgb(20, 20, 20);
	border: 0.5px solid rgb(104, 104, 104);
}

.codeInput:focus {
	border: 0.5px solid rgb(160, 160, 160);
}

.applyBtn {
	width: 100%;
	margin-top: 0.5rem;
}

.costCalculation {
	margin: 1rem 0;
}

.wrapper {
	margin: 4px 0;
}

.shippingCont {
	position: relative;
}

.infoBtn {
	margin: 0;
	padding: 0;
	border: none;

	position: absolute;
	top: -4px;
	left: 62px;
}

.infoBtn:hover {
	margin: 0;
	padding: 0;
	border: none;
}

.infoBtn > img {
	vertical-align: middle;
}

.freeShippingTooltip {
	width: 170px;
	padding: 6px;

	font-size: 0.8rem;
	background-color: rgb(20, 20, 20);
	border: 1px solid rgb(104, 104, 104);

	position: absolute;
	top: -53px;
	left: 73px;
}

.totalWrapper {
	padding: 1rem 0;
	border-top: 2px solid rgb(104, 104, 104);

	position: relative;
}

.totalLayout {
	font-family: 'DINNext bold';
	font-size: 1.2rem;
}

.creditsText {
	font-family: DINNext;
	font-size: 1rem;

	position: absolute;
	right: 0;
	top: 38px;
}

.checkoutBtn {
	margin: 1rem 0 0.5rem 0;
	padding: 1rem 0.5rem;

	font-family: 'DINNext bold';

	transition: all 0.5s ease;
}

.checkoutBtn:hover {
	background-color: rgb(0, 116, 52);
}

.continueLink {
	text-align: center;
	color: rgb(158, 158, 158);
	transition: all 0.3s ease;
}

.continueLink:hover {
	color: rgb(221, 221, 221);
}

@media (max-width: 1150px) {
	.cartSummaryCont {
		padding-left: 1rem;
	}
}

@media (max-width: 620px) {
	.cartSummaryCont {
		padding-left: 0;
		border-left: none;

		padding-top: 1rem;

		position: static;
	}
}
