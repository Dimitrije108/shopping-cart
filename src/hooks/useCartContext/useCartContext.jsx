import { useState, useEffect, createContext, useContext } from 'react';

const CartContext = createContext({
  cart: null,
	limitExceededMsg: "", 
  addToCart: () => {},
	removeFromCart: () => {},
});
// Shopping cart functionality passed on to all components 
// via context provider
export function useShoppingCart() {
	// Retrieve cart from storage if available
	const storedCart = JSON.parse(localStorage.getItem('cart'));
	const initCart = storedCart ? storedCart : [];

	const [cart, setCart] = useState(initCart);

	useEffect(() => {
		localStorage.setItem('cart', JSON.stringify(cart));
	}, [cart]);
	// Warning message for exceeding the ship quantity limit
	const limitExceededMsg = 
		"For orders larger than 20 ships per model please contact us directly!";
	// Add an item or adjust it's quantity
	const addToCart = (item) => {
		// check if the provided item is already inside the cart
		const isDuplicate = cart.find((ship) => ship.id === item.id);
		// handle duplicate items
		if (isDuplicate) {
			const totalQuantity = isDuplicate.quantity + item.quantity;
			// handle exceeding item quantity
			if (totalQuantity > 20) {
				return false;
			};
			// adjust quantity accordingly
			const addQuantity = cart.map((ship) => {
				if (ship.id === item.id) {
					return { ...ship, quantity: ship.quantity + item.quantity };
				}
				return ship;
			});
			setCart(addQuantity);
			return true;
		} else {
			// add unique item to the cart
			setCart([...cart, item]);
			return true;
		};
	};
	// Remove item instance from the cart
	const removeFromCart = (id) => {
		const removed = cart.filter((ship) => ship.id !== id);
		setCart(removed);
	};

	return { 
		cart, 
		limitExceededMsg, 
		addToCart, 
		removeFromCart,
	};
};

export const useCartContext = () => useContext(CartContext);

export default CartContext;