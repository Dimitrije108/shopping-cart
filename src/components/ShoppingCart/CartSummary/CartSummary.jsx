import { useState, useMemo } from "react";
import { Link } from "react-router-dom";
import { useCartContext } from "../../../hooks/useCartContext/useCartContext";
import formatNumber from "../../../formatNumber/formatNumber";
import styles from './CartSummary.module.css';

// TODO: make a contact form for "Contact for price" items
// TODO: introduce discount functionality
export default function CartSummary() {
	const [freeShippingTooltip, setFreeShippingTooltip] = useState(false);
	const { cart } = useCartContext();

	const toggleTooltip = () => {
		setFreeShippingTooltip((prev) => !prev);
	};
	// Calculate and memoize subtotal
	const subtotal = useMemo(() => {
		return cart.reduce((total, ship) => {
			return ship.price !== "Price on Request" 
				? total + (ship.quantity * Number(ship.price)) 
				: total;
		}, 0);
	}, [cart]);
	// Calculate shipping into total cost
	const shippingCost = subtotal < 500000 && cart.length > 0
		? 25000
		: 0;
	// Calculate shipping display
	const shipping = shippingCost === 0
		? "Free"
		: formatNumber(shippingCost);
	
	const discount = 0;

	const total = subtotal + shippingCost - discount;
	
	return (
		<div className={styles.cartSummaryCont}>
			<h2 className={styles.summaryHeader}>Summary</h2>
			<div className={styles.discountWrapper}>
				<div className={styles.discountLayout}>
					<p className={styles.discount}>Discount code</p>
					<input 
						className={styles.codeInput}
						placeholder={"Type in code..."} 
					/>
				</div>
				<button className={styles.applyBtn}>Apply</button>
			</div>
			<div className={styles.costCalculation}>
				<div className={styles.wrapper}>
					<p>Subtotal</p>
					<div>{formatNumber(subtotal)}</div>
				</div>
				<div className={styles.wrapper}>
					<div className={styles.shippingCont}>
						<p>Shipping</p>
						<button 
							className={styles.infoBtn}
							onClick={toggleTooltip}
						>
							<img 
								src="/src/assets/info-circle.svg" 
								alt="free shipping tooltip" 
								width={17} 
								height={17} 
							/>
						</button>
						{freeShippingTooltip && 
								<div className={styles.freeShippingTooltip}>
									Free shipping for orders larger than 500,000 credits!
								</div>
							}
					</div>
					<div>{shipping}</div>
				</div>
				<div className={styles.wrapper}>
					<p>Discount</p>
					<div>0</div>
				</div>
			</div>
			<div className={styles.totalWrapper}>
				<div className={styles.totalLayout}>
					<h3>Total</h3>
					<div  className={styles.totalPrice}>
						{formatNumber(total)}
					</div>
				</div>
				{total > 0 && <span className={styles.creditsText}>credits</span>}
			</div>
			<button className={styles.checkoutBtn}>Checkout</button>
			<Link to="/starships" className={styles.continueLink}>Continue shopping</Link>
		</div>
	)
};
