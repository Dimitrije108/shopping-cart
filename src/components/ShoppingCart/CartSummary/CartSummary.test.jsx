import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MemoryRouter } from "react-router-dom";
import CartSummary from "./CartSummary";
import { useCartContext } from "../../../hooks/useCartContext/useCartContext";

const mockCartItem = {
	id: "6429291f021f17e13fbc1db9",
	img: "https://lumiere-a.akamaihd.net/v1/images/millennium-falcon-main-tlj-a_7cf89d3a.jpeg",
	name: "Millennium Falcon",
	category: "starfighter",
	price: "100000",
	quantity: 1,
};

const mockCartItem2 = {
	id: "6429291f021f17e13fbc1e3c",
	img: "https://lumiere-a.akamaihd.net/v1/images/X-Wing-Fighter_47c7c342.jpeg",
	name: "X-wing Starfighter",
	category: "starfighter",
	price: "149999",
	quantity: 3,
};

const mockCart = [mockCartItem, mockCartItem2];

vi.mock("../../../hooks/useCartContext/useCartContext", () => ({
	useCartContext: vi.fn(() => ({
		cart: mockCart,
	}))
}));

describe("Cart Summary component", () => {
	it("cart summary snapshot test", () => {
		const cartSummary = render(
			<MemoryRouter>
				<CartSummary />
			</MemoryRouter>
		)

		expect(cartSummary).toMatchSnapshot();
	});

	describe("Renders tests", () => {
		it("renders summary heading", () => {
			render(
				<MemoryRouter>
					<CartSummary />
				</MemoryRouter>
			)
	
			const heading = screen.getByRole("heading", { name: "Summary" });
	
			expect(heading).toBeInTheDocument();
		});
	
		it("renders submit discount", () => {
			render(
				<MemoryRouter>
					<CartSummary />
				</MemoryRouter>
			)
			
			const para = screen.getByText("Discount code");
			const input = screen.getByPlaceholderText("Type in code...");
			const btn = screen.getByRole("button", { name: "Apply" });
	
			expect(para).toBeInTheDocument();
			expect(input).toBeInTheDocument();
			expect(btn).toBeInTheDocument();
		});
	
		it("renders subtotal content", () => {
			render(
				<MemoryRouter>
					<CartSummary />
				</MemoryRouter>
			)
			
			const para = screen.getByText("Subtotal");
			const subtotal = screen.getByTestId("subtotal");
	
			expect(para).toBeInTheDocument();
			expect(subtotal).toHaveTextContent("549,997");
		});

		it("renders shipping content", () => {
			useCartContext.mockReturnValueOnce({
				cart: [mockCartItem2],
			});

			render(
				<MemoryRouter>
					<CartSummary />
				</MemoryRouter>
			)
			
			const para = screen.getByText("Shipping");
			const infoBtn = screen.getByRole("button", { name: "free shipping tooltip" });
			const infoBtnIcon = screen.getByRole("img", { name: "free shipping tooltip" });
			const shippingPrice = screen.getByTestId("shipping");
	
			expect(para).toBeInTheDocument();
			expect(infoBtn).toBeInTheDocument();
			expect(infoBtnIcon).toHaveAttribute("src", "/src/assets/icons/info-circle.svg");
			expect(shippingPrice).toHaveTextContent("25,000");
		});

		it("renders discount content", () => {
			render(
				<MemoryRouter>
					<CartSummary />
				</MemoryRouter>
			)
			
			const para = screen.getByText("Discount");
			const discount = screen.getByTestId("discount");
	
			expect(para).toBeInTheDocument();
			expect(discount).toHaveTextContent("0");
		});

		it("renders total cost content", () => {
			render(
				<MemoryRouter>
					<CartSummary />
				</MemoryRouter>
			)
			
			const heading = screen.getByRole("heading", { name: "Total" });
			const total = screen.getByTestId("total");
			const credits = screen.getByText("credits");
	
			expect(heading).toBeInTheDocument();
			expect(total).toHaveTextContent("549,997");
			expect(credits).toBeInTheDocument();
		});

		it("renders total cost content", () => {
			render(
				<MemoryRouter>
					<CartSummary />
				</MemoryRouter>
			)
			
			const checkoutBtn = screen.getByRole("button", { name: "Checkout" });
			const continueShoppingLink = screen.getByRole(
				"link", { name: "Continue shopping" }
			);

			expect(checkoutBtn).toBeInTheDocument();
			expect(continueShoppingLink).toBeInTheDocument();
		});
	})

	describe("Logic tests", () => {
		it("tooltip toggles on/off after button click", async () => {
			const user = userEvent.setup();

			render(
				<MemoryRouter>
					<CartSummary />
				</MemoryRouter>
			)

			const tooltipBtn = screen.getByRole("img", { name: "free shipping tooltip" });
			
			await user.click(tooltipBtn);

			const tooltip = screen.getByText("Free shipping for orders larger than 500,000 credits!");
			expect(tooltip).toBeInTheDocument();

			await user.click(tooltipBtn);

			expect(tooltip).not.toBeInTheDocument();
		});

		it("subtotal calculates correctly", () => {
			render(
				<MemoryRouter>
					<CartSummary />
				</MemoryRouter>
			)
			
			const subtotal = screen.getByTestId("subtotal");
	
			expect(subtotal).toHaveTextContent("549,997");
		});

		it("shipping cost renders 'Free' for orders larger than set threshold", () => {
			render(
				<MemoryRouter>
					<CartSummary />
				</MemoryRouter>
			)
			
			const shippingPrice = screen.getByTestId("shipping");
	
			expect(shippingPrice).toHaveTextContent("Free");
		});

		it("shipping cost renders shipping price for orders smaller than set threshold", () => {
			useCartContext.mockReturnValueOnce({
				cart: [mockCartItem],
			});
			
			render(
				<MemoryRouter>
					<CartSummary />
				</MemoryRouter>
			)
			
			const shippingPrice = screen.getByTestId("shipping");
	
			expect(shippingPrice).toHaveTextContent("25,000");
		});
		
		it("total calculates correctly", () => {
			useCartContext.mockReturnValueOnce({
				cart: [mockCartItem2],
			});
			
			render(
				<MemoryRouter>
					<CartSummary />
				</MemoryRouter>
			)
			
			const total = screen.getByTestId("total");
	
			expect(total).toHaveTextContent("474,997");
		});
	})
})