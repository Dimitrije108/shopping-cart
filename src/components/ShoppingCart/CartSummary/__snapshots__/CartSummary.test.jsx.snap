// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cart Summary component > cart summary snapshot test 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_cartSummaryCont_00018c"
      >
        <h2
          class="_summaryHeader_00018c"
        >
          Summary
        </h2>
        <div
          class="_discountWrapper_00018c"
        >
          <div
            class="_discountLayout_00018c"
          >
            <p
              class="_discount_00018c"
            >
              Discount code
            </p>
            <input
              class="_codeInput_00018c"
              placeholder="Type in code..."
            />
          </div>
          <button
            class="_applyBtn_00018c"
          >
            Apply
          </button>
        </div>
        <div
          class="_costCalculation_00018c"
        >
          <div
            class="_wrapper_00018c"
          >
            <p>
              Subtotal
            </p>
            <div
              data-testid="subtotal"
            >
              549,997
            </div>
          </div>
          <div
            class="_wrapper_00018c"
          >
            <div
              class="_shippingCont_00018c"
            >
              <p>
                Shipping
              </p>
              <button
                class="_infoBtn_00018c"
              >
                <img
                  alt="free shipping tooltip"
                  height="17"
                  src="/src/assets/info-circle.svg"
                  width="17"
                />
              </button>
            </div>
            <div
              data-testid="shipping"
            >
              Free
            </div>
          </div>
          <div
            class="_wrapper_00018c"
          >
            <p>
              Discount
            </p>
            <div
              data-testid="discount"
            >
              0
            </div>
          </div>
        </div>
        <div
          class="_totalWrapper_00018c"
        >
          <div
            class="_totalLayout_00018c"
          >
            <h3>
              Total
            </h3>
            <div
              class="_totalPrice_00018c"
              data-testid="total"
            >
              549,997
            </div>
          </div>
          <span
            class="_creditsText_00018c"
          >
            credits
          </span>
        </div>
        <button
          class="_checkoutBtn_00018c"
        >
          Checkout
        </button>
        <a
          class="_continueLink_00018c"
          data-discover="true"
          href="/starships"
        >
          Continue shopping
        </a>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_cartSummaryCont_00018c"
    >
      <h2
        class="_summaryHeader_00018c"
      >
        Summary
      </h2>
      <div
        class="_discountWrapper_00018c"
      >
        <div
          class="_discountLayout_00018c"
        >
          <p
            class="_discount_00018c"
          >
            Discount code
          </p>
          <input
            class="_codeInput_00018c"
            placeholder="Type in code..."
          />
        </div>
        <button
          class="_applyBtn_00018c"
        >
          Apply
        </button>
      </div>
      <div
        class="_costCalculation_00018c"
      >
        <div
          class="_wrapper_00018c"
        >
          <p>
            Subtotal
          </p>
          <div
            data-testid="subtotal"
          >
            549,997
          </div>
        </div>
        <div
          class="_wrapper_00018c"
        >
          <div
            class="_shippingCont_00018c"
          >
            <p>
              Shipping
            </p>
            <button
              class="_infoBtn_00018c"
            >
              <img
                alt="free shipping tooltip"
                height="17"
                src="/src/assets/info-circle.svg"
                width="17"
              />
            </button>
          </div>
          <div
            data-testid="shipping"
          >
            Free
          </div>
        </div>
        <div
          class="_wrapper_00018c"
        >
          <p>
            Discount
          </p>
          <div
            data-testid="discount"
          >
            0
          </div>
        </div>
      </div>
      <div
        class="_totalWrapper_00018c"
      >
        <div
          class="_totalLayout_00018c"
        >
          <h3>
            Total
          </h3>
          <div
            class="_totalPrice_00018c"
            data-testid="total"
          >
            549,997
          </div>
        </div>
        <span
          class="_creditsText_00018c"
        >
          credits
        </span>
      </div>
      <button
        class="_checkoutBtn_00018c"
      >
        Checkout
      </button>
      <a
        class="_continueLink_00018c"
        data-discover="true"
        href="/starships"
      >
        Continue shopping
      </a>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
